test/junit/src/test/scala/strawman/collection/TupleZippedTest.scala

```diff
@@ -0,0 +1,179 @@
+    val res: List[(Int, Int, String)] = zipped3.map((a, b, c) => (a, b, c))
+
+    assertEquals(List((1, 1, "a"), (2, 2, "b"), (3, 3, "c")), res)
+  }
+
+  @Test
+  def tuple3Zipped_flatMap(): Unit = {
+    val res: List[(Int, Int, String)] = zipped3.flatMap((a, b, c) => List((a, b, c)))
+
+    assertEquals(List((1, 1, "a"), (2, 2, "b"), (3, 3, "c")), res)
+  }
+
+  @Test
+  def tuple3Zipped_filter(): Unit = {
+    val res: List[(Int, Int, String)] = zipped3.filter((a, _, _) => a % 2 != 0)
+
+    assertEquals(List((1, 1, "a"), (3, 3, "c")), res)
+  }
+
+  @Test
+  def tuple3Zipped_exists(): Unit = {
+    assertTrue(zipped3.exists((a, b, _) => a + b > 5))
+    assertFalse(zipped3.exists((a, b, _) => a + b < 0))
+  }
+
+  @Test
+  def tuple3Zipped_forall(): Unit = {
+    assertTrue(zipped3.forall((a, b, _) => (a + b) % 2 == 0))
+    assertFalse(zipped3.forall((a, b, _) => a + b > 5))
+  }
+
+  @Test
+  def tuple3Zipped_foreach(): Unit = {
+    var res = ""
+    zipped3.foreach((a, b, c) => res += s"[$a,$b,$c]")
+
+    assertEquals("[1,1,a][2,2,b][3,3,c]", res)
+  }
+
+  @Test
+  def tuple3Zipped_toIterable(): Unit = {
+    val iter: Iterable[(Int, Int, String)] = zipped3
+
+    assertEquals(List((1, 1, "a"), (2, 2, "b"), (3, 3, "c")), iter.to(List))
+  }
+
+  @Test
+  def tuple3Zipped_empty(): Unit = {
+    assertTrue(zipped2.lazyZip(List.empty).isEmpty)
+  }
+
+  @Test
+  def tuple4Zipped_map(): Unit = {
+    val res: List[(Int, Int, String, Boolean)] = zipped4.map((a, b, c, d) => (a, b, c, d))
+
+    assertEquals(List((1, 1, "a", true), (2, 2, "b", false), (3, 3, "c", true)), res)
+  }
+
+  @Test
+  def tuple4Zipped_flatMap(): Unit = {
+    val res: List[(Int, Int, String, Boolean)] = zipped4.flatMap((a, b, c, d) => List((a, b, c, d)))
+
+    assertEquals(List((1, 1, "a", true), (2, 2, "b", false), (3, 3, "c", true)), res)
+  }
+
+  @Test
+  def tuple4Zipped_filter(): Unit = {
+    val res: List[(Int, Int, String, Boolean)] = zipped4.filter((_, _, _, d) => d)
+
+    assertEquals(List((1, 1, "a", true), (3, 3, "c", true)), res)
+  }
+
+  @Test
+  def tuple4Zipped_exists(): Unit = {
+    assertTrue(zipped4.exists((a, b, c, d) => a + b > 5 && !c.isEmpty && d))
+    assertFalse(zipped4.exists((a, b, c, d) => a + b > 5 && !c.isEmpty && !d))
+  }
+
+  @Test
+  def tuple4Zipped_forall(): Unit = {
+    assertTrue(zipped4.forall((a, b, _, _) => (a + b) % 2 == 0))
+    assertFalse(zipped4.forall((a, b, _, d) => a + b > 0 && d))
+  }
+
+  @Test
+  def tuple4Zipped_foreach(): Unit = {
+    var res = ""
+    zipped4.foreach((a, b, c, d) => res += s"[$a,$b,$c,$d]")
+
+    assertEquals("[1,1,a,true][2,2,b,false][3,3,c,true]", res)
+  }
+
+  @Test
+  def tuple4Zipped_toIterable(): Unit = {
+    val iter: Iterable[(Int, Int, String, Boolean)] = zipped4
+
+    assertEquals(List((1, 1, "a", true), (2, 2, "b", false), (3, 3, "c", true)), iter.to(List))
+  }
+
+  @Test
+  def tuple4Zipped_empty(): Unit = {
+    assertTrue(zipped3.lazyZip(List.empty).isEmpty)
+  }
+}
```

Comment:
Can you also check that `TreeSet(1, 2, 3).lazyZip(List(4, 5, 6)).map((x, y) => x + y)` returns a `TreeSet`?

And that `TreeMap(1 -> "foo", 2 -> "bar").lazyZip(List("baz", "bah")).map { case ((k, v), s) => k -> (v ++ s) }` also returns a `TreeMap`?

Reply:
it does not work well with `TreeMap` :(

Reply:
The issue here doesn't only concern `TreeMap` but `Map`s in general. I guess the reason is the same for having overloaded methods in `Map` (e.g. `map`, `flatMap`, `concat`).

Having a more specific implicit conversion in the `Map` companion object may address this:

```
implicit class LazyZipOps[K, V, CC1[K, V] <: Map[K, V]](`this`: CC1[K, V]) {
  def lazyZip[B, CC2[B] <: Iterable[B]](that: CC2[B]): Tuple2Zipped[(K, V), CC1[K, V], B, CC2[B]] = new Tuple2Zipped(`this`, that)
}
```

Reply:
Yes, that’s a good idea.

Also, I think we don’t need to keep track of the precise type of the zipped collections: we just need `C1` (to drive the type of the collection resulting from a transformation operation applied to the zipped collections), but we never need `C2`, `C3` or `C4`.

So, I think we could simplify things as follows:

~~~ scala
// TupleZipped.scala
final class Tuple2Zipped[El1, El2, C1 <: Iterable[El1]](coll1: C1, coll2: Iterable[El2]) { … }
final class Tuple3Zipped[El1, El2, El3, C1 <: Iterable[El1]](coll1: C1, coll2: Iterable[El2], coll3: Iterable[El3]) { … }
// same for Tuple4Zipped

class LazyZipOps[A, C1 <: Iterable[A]](`this`: C1) {
  def lazyZip[B](that: Iterable[B]): Tuple2Zipped[A, B, C1] = new Tuple2Zipped[A, B, C1](`this`, that)
}

// Iterable.scala
object Iterable {
  implicit def toLazyZipOps[CC[X] <: Iterable[X], A](it: CC[A]): LazyZipOps[A, CC[A]] = new LazyZipOps(it)
}

// Map.scala
object Map {
  implicit def toLazyZipOps[CC[X, Y] <: Iterable[(X, Y)], K, V](it: CC[K, V]): LazyZipOps[(K, V), CC[K, V]] = new LazyZipOps(it)
}
~~~

Reply:
yeah, I was considering doing that too. The collection types were needed in previous iterations when `filter` was equivalent to `(xs zip ys).filter(f.tupled).unzip`, but now I can get rid of them